packages/rest/src/controllers/types.ts

```diff
@@ -175,20 +172,55 @@ export interface ConnectionInvitationSchema {
   imageUrl?: string
 }
 
-export interface RequestProofOptions extends ProofRequestConfig {
+export interface RequestProofOptionsProofRequestRestriction {
```

Comment:
I think we should match this with the API as we have in AFJ, and not create custom properties here.


The API in AFJ is:

```
{
  connectionId: string
  protocolVersion: string // 'v1' or 'v2'
  proofFormats: {
   <dynamic interface>
  }
}
```

You can use the whole `RequestProofOptions` type from AFJ to get the correct types and input interface.


You can define the proof protocols type once like this:

```
type ProofProtocols = [V1ProofProtocol, V2ProofProtocol<[LegacyIndyProofFormatService, AnonCredsProofFormatService]>]
```

and then in the rest api calls you can use as the type:

```
RequestProofOptions<ProofProtocols>
```

One thing you may run into is that the generated swagger is invalid, and this is due to some bugs in tsoa: https://github.com/lukeautry/tsoa/issues?q=is%3Aopen+is%3Aissue+author%3A%40me+sort%3Aupdated-desc

Maybe this is resolved by updating to the latest version, but if still the case, you need to define the top level object yourself, but can still infer the format types: 

See this as an example: https://github.com/hyperledger/aries-framework-javascript-ext/blob/main/packages/rest/src/controllers/types.ts#L125


For the request proof interface this would look something like:
```ts
// Define once, can be reused
type ProofFormats = [LegacyIndyProofFormat, AnonCredsProofFormat]

interface RestRequestProofOptions {
  connectionId: string
  autoAcceptProof?: AutoAcceptProof
  comment?: string
  proofFormats: ProofFormatPayload<ProofFormats, 'createRequest'>
}
```

It may seem a bit complex, but it helps in keeping the API of the rest package consistent with what we have in the JS api.

Let me know if you have any questions

Reply:
In general, looking at how it's done in the credentials controller is a good example, as this one was already updated to match the new api from 0.3.x which overhauled the credentials api but not yet the proofs api

Reply:
I understand the reasonning here but there are some problems with `attr::${string}::marker` and `attr::${string}::value` specifically. Notably tsoa does not deal well with types containing patterned keys like:

```ts
type Foo {
  [key: `attr::${string}::marker`]: '1' | '0'
  [key: `attr::${string}::value`]: string,
}
```

including types like this in a body will result in tsoa throwing an internal `GenerateMetadataError` when we build the spec and routes. (I'll see if I can find the issue that discussed this but I can't find it to hand)

This essentially leaves us with two choices:

1. match the API using just a `{ [key: string]: unknown }` and then parse out and validate the `marker` and `value`entries by hand
2. introduce new properties that represent these concepts more neatly within tsoa

I understand the value of consistency but choosing option 1. will lead to a poorly documented OpenAPI type and potentially buggy and error prone parseing logic being done by hand, which feels messy to say the least.

I don't really like the idea of the API being dictated by an issue with tooling, bu tin this instance I felt it was the lesser evil

Thoughts?

Reply:
I agree with you here. Maybe we can use the structure of the class in this case? So it will look something like:

```
restrictions: [
        {
          attributeValues: {
            test_prop: 'test_value',
            test_prop2: 'test_value2',
          },
          attributeMarkers: {
            test_prop: true,
            test_prop2: true,
          },
        },
      ],
```